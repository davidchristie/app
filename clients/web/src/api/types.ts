export interface Api {
  useAuthorize(): (providerId: string) => void;
  useSession: UseQuery<Session>;
}

export interface Session {
  user: User | null;
}

export interface QueryResult<Data> {
  data?: Data;
  error?: Error;
  loading: boolean;
}

export interface UseQuery<Data> {
  (): QueryResult<Data>;
}

export interface User {
  id: string;
  name: string;
  email: string;
}
