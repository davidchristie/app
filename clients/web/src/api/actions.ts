export function authorize(providerId: string): void {
  window.location.href = `/api/v1/auth/${providerId}/authorize`;
}
