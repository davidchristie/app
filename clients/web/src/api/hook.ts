import { useContext } from "react";
import { ApiContext } from "./context";
import { Api } from "./types";

export function useApi(): Api {
  const api = useContext(ApiContext);
  if (!api) {
    throw new Error("API context not found");
  }
  return api;
}
