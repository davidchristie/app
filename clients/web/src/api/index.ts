export * from "./hook";
export * from "./mock";
export * from "./provider";
export * from "./types";
