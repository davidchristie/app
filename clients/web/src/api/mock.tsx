import { ReactNode } from "react";
import { ApiContext } from "./context";
import { Api, Session, UseQuery } from "./types";

export interface MockApiData {
  authorize?: () => void;
  session?: Session | Error;
}

export interface MockApiProviderProps {
  children?: ReactNode;
  data?: MockApiData;
}

function createQueryHook<Data>(
  value: Data | Error | undefined
): UseQuery<Data> {
  return () => {
    if (!value) {
      return {
        loading: true,
      };
    } else if (value instanceof Error) {
      return {
        error: value,
        loading: false,
      };
    }
    return {
      data: value,
      loading: false,
    };
  };
}

export function MockApiProvider({
  children,
  data = {},
}: MockApiProviderProps): JSX.Element {
  const api: Api = {
    useAuthorize: () => data.authorize ?? jest.fn(),
    useSession: createQueryHook(data.session),
  };
  return <ApiContext.Provider value={api}>{children}</ApiContext.Provider>;
}
