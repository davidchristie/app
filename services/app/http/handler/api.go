package handler

import (
	"net/http"

	"github.com/davidchristie/app/services/app/auth"
	"github.com/davidchristie/app/services/app/http/middleware"
	"github.com/go-chi/chi/v5"
)

func apiHandler(auth auth.Auth) http.Handler {
	r := chi.NewRouter()
	r.Use(middleware.Session(auth))
	r.Mount("/v1", restV1Handler(auth))
	return r
}

func restV1Handler(auth auth.Auth) http.Handler {
	r := chi.NewRouter()
	r.Get("/auth/session", Session())
	r.Get("/auth/{providerID}/authorize", Authorize(auth))
	r.Get("/auth/{providerID}/callback", OAuthCallback(auth))
	r.Get("/auth/signout", SignOut())
	return r
}
