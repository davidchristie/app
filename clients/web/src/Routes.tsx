import { Route, RouteProps, Switch } from "react-router-dom";
import { useApi } from "./api";
import { HomePage } from "./pages/HomePage";
import { NotFoundPage } from "./pages/NotFoundPage";
import { SignInPage } from "./pages/SignInPage";

const authenticatedRoutes: RouteProps[] = [
  {
    exact: true,
    path: "/",
    component: HomePage,
  },
  {
    path: "*",
    component: NotFoundPage,
  },
];

const unauthenticatedRoutes: RouteProps[] = [
  {
    exact: true,
    path: "/",
    component: HomePage,
  },
  {
    exact: true,
    path: "/signin",
    component: SignInPage,
  },
  {
    path: "*",
    component: NotFoundPage,
  },
];

export function Routes(): JSX.Element {
  const { useSession } = useApi();
  const session = useSession();
  if (session.data === undefined) {
    return <>Loading...</>; // TODO
  }
  const routes =
    session.data.user === null ? unauthenticatedRoutes : authenticatedRoutes;
  return (
    <Switch>
      {routes.map((route, index) => (
        <Route {...route} key={index} />
      ))}
    </Switch>
  );
}
