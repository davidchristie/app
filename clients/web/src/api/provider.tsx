import { ReactNode } from "react";
import useSWR from "swr";
import { authorize } from "./actions";
import { ApiContext } from "./context";
import { Api, UseQuery } from "./types";

export interface ApiProviderProps {
  children?: ReactNode;
}

function createQueryHook<Data>(url: string): UseQuery<Data> {
  return () => {
    const { data, error } = useSWR(url);
    const loading = !error && !data;
    return { data, error, loading };
  };
}

export function ApiProvider({ children }: ApiProviderProps): JSX.Element {
  const api: Api = {
    useAuthorize: () => authorize,
    useSession: createQueryHook("/api/v1/auth/session"),
  };
  return <ApiContext.Provider value={api}>{children}</ApiContext.Provider>;
}
