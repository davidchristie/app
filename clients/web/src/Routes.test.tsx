import { StaticRouter } from "react-router-dom";
import { MockApiProvider } from "./api";
import { render, screen } from "./testing";
import { signedInSession, signedOutSession } from "./testing/data";
import { Routes } from "./Routes";

describe("Routes", () => {
  describe("loading", () => {
    it("renders correctly", () => {
      const { container } = render(<Routes />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("signed in", () => {
    const routes: { expectedTestId: string; url: string }[] = [
      {
        expectedTestId: "HomePage",
        url: "/",
      },
    ];

    routes.forEach(({ expectedTestId, url }) => {
      it(url, () => {
        render(
          <StaticRouter location={url}>
            <MockApiProvider data={{ session: signedInSession }}>
              <Routes />
            </MockApiProvider>
          </StaticRouter>
        );
        expect(screen.getAllByTestId(expectedTestId)).toBeDefined();
      });
    });
  });

  describe("signed out", () => {
    const routes: { expectedTestId: string; url: string }[] = [
      {
        expectedTestId: "HomePage",
        url: "/",
      },
      {
        expectedTestId: "SignInPage",
        url: "/signin",
      },
    ];

    routes.forEach(({ expectedTestId, url }) => {
      it(url, () => {
        render(
          <StaticRouter location={url}>
            <MockApiProvider data={{ session: signedOutSession }}>
              <Routes />
            </MockApiProvider>
          </StaticRouter>
        );
        expect(screen.getAllByTestId(expectedTestId)).toBeDefined();
      });
    });
  });
});
