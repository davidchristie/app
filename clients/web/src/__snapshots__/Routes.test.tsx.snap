// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes loading renders correctly 1`] = `Loading...`;
